cmd/ulimit_unix.go

```diff
@@ -15,37 +15,87 @@ func init() {
 	fileDescriptorF = checkAndSetUlimitUnix
 }
 
+// In this entire file, by 'ulimit' we mean the "num open files ulimit" (RLIMIT_NOFILE)
 func checkAndSetUlimitUnix() {
 	logger := zerolog.Logger
-	if err := setUlimit(ulimitUnix); err != nil {
-		logger.Err(fmt.Errorf("error setting ulimit: %w", err))
+
+	rlimit, err := getUlimit()
+	if err != nil {
+		logger.Err(err).Msg("failed getting ulimit")
+		return
+	}
+	logger.Info().Uint64("hard_ulimit", rlimit.Max).Uint64("soft_ulimit", rlimit.Cur).Msg("limits (before adjustment)")
+
+	if err := setUlimit(); err != nil {
+		logger.Err(err).Msg("failed setting ulimit")
+	}
+
+	rlimit, err = getUlimit()
+	if err != nil {
+		logger.Err(err).Msg("failed getting ulimit")
+		return
+	}
+	logger.Info().Uint64("hard_ulimit", rlimit.Max).Uint64("soft_ulimit", rlimit.Cur).Msg("limits (after adjustment)")
+}
+
+func setUlimit() error {
+	logger := zerolog.Logger
+
+	// Setting the hard ulimit is very likely to fail (usually requires root or even more arcane permissions).
+	// If it fails we just log at "INFO" level.
+	if err := setHardUlimit(); err != nil {
+		// Just an info log here, since it's really not actionable.
+		logger.Info().AnErr("err", err).Msg("failed setting hard ulimit")
 	}
+
+	return setSoftUlimit()
 }
 
-func setUlimit(ulimit uint64) error {
+func setHardUlimit() error {
 	logger := zerolog.Logger
-	rLimit, err := GetUlimit()
+	rLimit, err := getUlimit()
 	if err != nil {
 		return fmt.Errorf("error getting ulimit: %w", err)
 	}
 
-	logger.Info().Uint64("hard_ulimit", rLimit.Max).Uint64("soft_ulimit", rLimit.Cur).Msg("limits (before adjustment)")
+	if rLimit.Max < ulimitUnix {
+		logger.Info().Uint64("previous_ulimit", rLimit.Max).Uint64("new_ulimit", ulimitUnix).Msg("adjusting hard ulimit")
+		rLimit.Max = ulimitUnix
+		return syscall.Setrlimit(syscall.RLIMIT_NOFILE, &rLimit)
+	}
+
+	return nil
+}
+
+func setSoftUlimit() error {
+	logger := zerolog.Logger
 
-	if rLimit.Max < ulimit {
-		logger.Info().Uint64("previous_ulimit", rLimit.Max).Uint64("new_ulimit", ulimit).Msg("adjusting max ulimit")
-		rLimit.Max = ulimit
+	rlimit, err := getUlimit()
```

Comment:
[non-blocking] Similar comment about avoiding the extra call. We can have `setHardUlimit` return the new hard limit and use that value here, WDYT?

Reply:
Leaving as-is for now, but will add the logging.
[Only the syscall in setHardLimit can be avoided - the syscall in setSoftLimit needs to happen because it has no way of knowing if setHardLimit succeeded or not].
[Because only one of them can be avoided, I'd prefer to have a cleaner interface and consistency in code and behaviour between them]...